awx_collection/plugins/modules/tower_job_template.py

```diff
@@ -12,410 +13,430 @@
+        name=dict( required=True, type='str',),
+        new_name=dict(required=False, type='str'),
+        use_fact_cache=dict( required=False, type='bool', default='False', aliases=['fact_caching_enabled']),
+        extra_vars=dict( required=False, type='dict', default={}, aliases=['ask_extra_vars']),
+        verbosity=dict( required=False, type='str', choices=[ '0', '1', '2', '3', '4', '5', ], default='0',),
+        ask_credential_on_launch=dict( required=False, type='bool', default='False', aliases=['ask_credential']),
+        project=dict( required=False, type='str',),
+        webhook_service=dict( required=False, type='str', choices=[ 'github', 'gitlab', ],),
+        timeout=dict( required=False, type='int', default='0',),
+        ask_scm_branch_on_launch=dict( required=False, type='bool', default='False',),
         state=dict(choices=['present', 'absent'], default='present'),
+        credentials=dict(required=False, type='list', default=[])
+#SURVEY_SPEC
     )
 
-    module = TowerModule(
-        argument_spec=argument_spec,
-        supports_check_mode=True,
-        mutually_exclusive=[
-            ('credential', 'credentials'),
-            ('vault_credential', 'credentials'),
-            ('extra_vars_path', 'extra_vars'),
-        ]
-    )
-
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    # Extract our parameters
+    scm_branch = module.params.get('scm_branch')
+    if len(scm_branch) > 1024:
+        module.fail_msg(msg="The value for scm_branch can not be longer than 1024")
+    ask_tags_on_launch = module.params.get('ask_tags_on_launch')
+    job_type = module.params.get('job_type')
+    skip_tags = module.params.get('skip_tags')
+    if len(skip_tags) > 1024:
+        module.fail_msg(msg="The value for skip_tags can not be longer than 1024")
+    ask_verbosity_on_launch = module.params.get('ask_verbosity_on_launch')
+    playbook = module.params.get('playbook')
+    if len(playbook) > 1024:
+        module.fail_msg(msg="The value for playbook can not be longer than 1024")
+    survey_enabled = module.params.get('survey_enabled')
+    description = module.params.get('description')
+    ask_diff_mode_on_launch = module.params.get('ask_diff_mode_on_launch')
+    job_tags = module.params.get('job_tags')
+    if len(job_tags) > 1024:
+        module.fail_msg(msg="The value for job_tags can not be longer than 1024")
+    allow_simultaneous = module.params.get('allow_simultaneous')
+    diff_mode = module.params.get('diff_mode')
+    ask_inventory_on_launch = module.params.get('ask_inventory_on_launch')
+    inventory = module.params.get('inventory')
+    limit = module.params.get('limit')
+    forks = module.params.get('forks')
+    if forks < 0:
+        module.fail_msg(msg="The value for forks can not be less than 0")
+    if forks > 2147483647:
+        module.fail_msg(msg="The value for forks can not be larger than 2147483647")
```

Comment:
It seems like the API validation should handle this just fine? I can take all the work you did here enumerating these and just exercise them in a test.

Reply:
This was all auto-generated from the API itself. I'm on the fence of do we let the API respond to things like this or just auto-enforce it within the module.